File: scripts/legoset.js
import { LegoService } from "./legoservice.js";

class LegoSet {
    Element;

    columns = ["setid", "name", "price", "originalprice", "discount", "retiring", "new"];

    constructor(core) {
        this['tracked'] = core['tracked'];
        this['have'] = core['have'];

        for (let current of this.columns) {
            this[current] = core[current];
        }
    }

    CreateRow() {
        this.Element = document.createElement('tr');
        this.Element.dataset.hasData = "yes";

        for (let current of ['have', 'tracked']) {
            let cell = document.createElement('td');
            let check = document.createElement('input');
            check['type'] = 'checkbox';
            check['checked'] = this[current];
            check.dataset.checkType = current;
            check.addEventListener('change', this.ChangeTrack.bind(this));
            cell.appendChild(check);
            this.Element.append(cell);
        }

        for (let current of this.columns) {
            let cell = document.createElement('td');
            if (current == "name")
            {
                let nameSpan = document.createElement('span');
                nameSpan.innerText = this[current];
                cell.appendChild(nameSpan);

                let legoAnchor = document.createElement('a');
                legoAnchor['href'] = 'https://lego.com/en-us/product/' + this['setid'];
                legoAnchor['target'] = '_blank';
                legoAnchor.innerText = '(Lego)';
                cell.appendChild(legoAnchor);

                let blAnchor = document.createElement('a');
                blAnchor['href'] = 'https://www.bricklink.com/v2/catalog/catalogitem.page?S=' + this['setid'];
                blAnchor['target'] = '_blank';
                blAnchor.innerText = '(BL)';
                cell.appendChild(blAnchor);

                let bsAnchor = document.createElement('a');
                bsAnchor['href'] = 'https://www.brickset.com/sets/' + this['setid'];
                bsAnchor['target'] = '_blank';
                bsAnchor.innerText = '(BS)';
                cell.appendChild(bsAnchor);
            }
            else
            {
                let string = this[current];
                if (current == 'retiring' || current == 'new')
                {
                    cell['style'] = 'text-align: center;';
                    if (this[current] == '1')
                        string = '\u2705';
                    else
                        string = '';
                }
                else if (current == 'discount')
                {
                    cell.classList.add('pricecell');
                    if (this[current] > 0)
                        string = (this[current] * 100) + '%';
                    else
                        string = '';
                }
                else if (current == 'price' || current == 'originalprice')
                {
                    cell.classList.add('pricecell');
                    string = '$' + this[current];
                }
                cell.appendChild(document.createTextNode(string));
            }
            this.Element.appendChild(cell);
        }
    }

    ChangeTrack(event) {
        this[event.target.dataset.checkType] = event.target.checked;
        LegoService.updateSet(
            this['setid'],
            this['have'],
            this['tracked'], this.Saved.bind(this));
    }

    Saved() {
        
    }
}

export { LegoSet };

File: api/models/legoSet.py
class LegoSet:
    def __init__(self, dataDict=None):
        if dataDict is not None:
            self.importData(dataDict)
        else:
            self.name = None
            self.categories = []
            self.originalPrice = None
            self.price = None
            self.discount = 0
            self.setid = None
            self.retiring = False
            self.new = False
            self.modified = None
            self.cancheck = True

    def __repr__(self):
        desc = f'{self.setid} - {self.name}: {self.price}'
        if self.discount != 0:
            desc = f'{desc} ({self.discount:%} {self.originalPrice})'
        if self.retiring:
            desc = f'{desc} (Retiring)'
        if self.new:
            desc = f'{desc} (New)'
        return desc

    def __eq__(self, other):
        if (other is None):
            return False
        return (
            self.name == other.name
            and self.price == other.price
            and self.originalPrice == other.originalPrice
            and self.discount == other.discount
            and self.setid == other.setid
            and self.new == other.new
            and self.retiring == other.retiring
        )

    def importData(self, dataDict):
        self.setid = dataDict['setid'] if 'setid' in dataDict else None
        self.originalPrice = dataDict['originalprice'] \
            if 'originalprice' in dataDict else None
        self.price = dataDict['price'] if 'price' in dataDict else None
        self.discount = dataDict['discount'] if 'discount' in dataDict else 0
        self.name = dataDict['name'] if 'name' in dataDict else None
        self.new = dataDict['new'] if 'new' in dataDict else False
        self.retiring = dataDict['retiring'] \
            if 'retiring' in dataDict else False
        self.cancheck = dataDict['cancheck'] \
            if 'cancheck' in dataDict else False
        self.modified = dataDict['modified'] \
            if 'modified' in dataDict else None
        self.categories = []

    def toJson(self):
        return f'{{ "name": "{self.name}",' \
            f'"price": {self.price or self.originalPrice},' \
            f'"originalprice": {self.originalPrice},' \
            f'"retiring": {"true" if self.retiring else "false"},' \
            f'"new": {"true" if self.new else "false"},' \
            f'"discount": {self.discount},' \
            f'"modified": "{self.modified}",' \
            f'"cancheck": {"true" if self.cancheck else "false"},' \
            f'"setid": {self.setid} }}'
